File: internal/handler/moderation.go
package handler

import (
	"context"
	"log/slog"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/fireynis/the-bell/internal/domain"
	"github.com/fireynis/the-bell/internal/middleware"
	"github.com/fireynis/the-bell/internal/service"
)

// ModerationHandler handles HTTP requests for moderation action operations.
type ModerationHandler struct {
	actions *service.ModerationActionService
}

// NewModerationHandler creates a ModerationHandler.
func NewModerationHandler(actions *service.ModerationActionService) *ModerationHandler {
	return &ModerationHandler{actions: actions}
}

type takeActionRequest struct {
	TargetUserID    string `json:"target_user_id"`
	ActionType      string `json:"action_type"`
	Severity        int    `json:"severity"`
	Reason          string `json:"reason"`
	DurationSeconds *int64 `json:"duration_seconds"`
}

// takeActionOutcome maps a TakeAction result onto the response status.
//
// An action that was persisted before its penalty propagation or enforcement
// failed is still a created action: reporting a failure would tell the
// moderator to retry and duplicate it. Those cases report 201 with partial
// set, so the caller can log the failure without failing the request.
func takeActionOutcome(result *service.TakeActionResult, err error) (status int, partial bool) {
	if err == nil {
		return http.StatusCreated, false
	}
	if result != nil && result.Action != nil {
		return http.StatusCreated, true
	}
	status, _ = statusForError(err)
	return status, false
}

// canQueryByModerator reports whether u may list the actions taken BY a
// moderator. That view exposes which moderator handled which case, so it is
// restricted to the council.
//
// The service refuses the same request for itself, and this is the early
// rejection rather than the only one: it is what gives the refusal this
// handler's error shape instead of the generic forbidden mapping.
func canQueryByModerator(u *domain.User) bool {
	return u != nil && u.IsCouncil()
}

// TakeAction handles POST /api/v1/moderation/actions.
func (h *ModerationHandler) TakeAction(w http.ResponseWriter, r *http.Request) {
	user, ok := middleware.UserFromContext(r.Context())
	if !ok {
		Error(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	var req takeActionRequest
	if err := Decode(r, &req); err != nil {
		Error(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := h.actions.TakeAction(
		r.Context(),
		user.ID,
		req.TargetUserID,
		domain.ActionType(req.ActionType),
		req.Severity,
		req.Reason,
		req.DurationSeconds,
	)
	status, partial := takeActionOutcome(result, err)
	if partial {
		// Either enforcement or penalty propagation failed; the joined error
		// says which.
		slog.Warn("moderation action created but follow-up failed",
			"action_id", result.Action.ID,
			"error", err,
		)
	} else if err != nil {
		serviceError(w, err)
		return
	}

	JSON(w, status, result)
}

// restrictionEndpoint is one time-boxed restriction as the HTTP layer sees it:
// how to read its status, how to lift it, and the response shape that names it.
//
// The service already reduced the mute and the suspension to a single
// restriction descriptor, because the two differ in nothing but the column they
// read. This is that reduction one layer up. The four routes were four bodies
// that agreed line for line on the session check, the URL parameter, the error
// mapping and the status codes, and nothing but review kept them agreeing —
// a 403 or a 204 that drifted on one of the pair would be a real difference in
// the API for no reason anybody chose.
//
// The response shapes stay distinct, and that is the one thing not shared: the
// JSON keys `muted_until` and `suspended_until` are the API, so each endpoint
// supplies its own body rather than a common one with a renamed field.
type restrictionEndpoint struct {
	// status and lift are method expressions on the service, so the pairing of
	// a status read with its matching lift is made once here rather than at
	// each of the four call sites.
	status func(*service.ModerationActionService, context.Context, *domain.User, string) (*time.Time, error)
	lift   func(*service.ModerationActionService, context.Context, *domain.User, string) error
	// body builds the status response from a formatted timestamp, empty when
	// the restriction is not in force.
	body func(until string) any
}

// muteStatusResponse is what a moderator may learn about someone else's mute:
// when it ends, and nothing more.
//
// MutedUntil is omitted entirely when the user is not muted, so the field's
// presence is the answer — the same rule ownProfileResponse uses, so a client
// reads both shapes the same way. It is also why this is an object rather than
// a bare timestamp.
type muteStatusResponse struct {
	MutedUntil string `json:"muted_until,omitempty"`
}

// suspensionStatusResponse is the suspension's counterpart to
// muteStatusResponse, field-for-field: the expiry when one is in force, and an
// empty object otherwise.
type suspensionStatusResponse struct {
	SuspendedUntil string `json:"suspended_until,omitempty"`
}

var (
	muteEndpoint = restrictionEndpoint{
		status: (*service.ModerationActionService).MuteStatus,
		lift:   (*service.ModerationActionService).LiftMute,
		body:   func(until string) any { return muteStatusResponse{MutedUntil: until} },
	}

	suspensionEndpoint = restrictionEndpoint{
		status: (*service.ModerationActionService).SuspensionStatus,
		lift:   (*service.ModerationActionService).LiftSuspension,
		body:   func(until string) any { return suspensionStatusResponse{SuspendedUntil: until} },
	}
)

// restrictionStatus answers a status query for either restriction.
//
// The service re-checks the moderator role; the route guard is the early
// rejection. Neither status route refuses a self-query, unlike the lifts.
func (h *ModerationHandler) restrictionStatus(w http.ResponseWriter, r *http.Request, e restrictionEndpoint) {
	user, ok := middleware.UserFromContext(r.Context())
	if !ok {
		Error(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	until, err := e.status(h.actions, r.Context(), user, chi.URLParam(r, "user_id"))
	if err != nil {
		serviceError(w, err)
		return
	}

	var formatted string
	if until != nil {
		formatted = until.Format(timestampFormat)
	}
	JSON(w, http.StatusOK, e.body(formatted))
}

// liftRestriction lifts either restriction.
//
// It answers 204 with no body, including for a user who was not restricted. The
// service treats that as done rather than as an error, the same as removing a
// reaction that was never left, and the status has to say the same thing or the
// endpoint is lying about what it did.
func (h *ModerationHandler) liftRestriction(w http.ResponseWriter, r *http.Request, e restrictionEndpoint) {
	user, ok := middleware.UserFromContext(r.Context())
	if !ok {
		Error(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	// The service re-checks the role; the route guard is the early rejection.
	if err := e.lift(h.actions, r.Context(), user, chi.URLParam(r, "user_id")); err != nil {
		serviceError(w, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// MuteStatus handles GET /api/v1/moderation/users/{user_id}/mute.
//
// This is the only response outside a caller's own profile that carries
// muted_until, and it lives behind the moderator guard for the reason the
// public profile does not carry it at all: a mute is between the user and the
// moderators, and this is the moderators' side of that. Putting it on
// GET /users/{id} would publish it to everyone, that route not even being
// authenticated.
//
// Without it a moderator's only clue would be a past mute action in the audit
// trail, which stays exactly as written after the mute is lifted and would go
// on reporting a mute that no longer exists.
func (h *ModerationHandler) MuteStatus(w http.ResponseWriter, r *http.Request) {
	h.restrictionStatus(w, r, muteEndpoint)
}

// LiftMute handles DELETE /api/v1/moderation/users/{user_id}/mute.
//
// A DELETE on the mute rather than a POST of an "unmute" action, because that
// is what it is: the mute is a value on the user (muted_until), and this
// removes it. Modelling it as an action would imply a moderation_actions row,
// which LiftMute deliberately does not write — see its doc comment.
//
// It answers 204 with no body, including for a user who was not muted; see
// liftRestriction.
func (h *ModerationHandler) LiftMute(w http.ResponseWriter, r *http.Request) {
	h.liftRestriction(w, r, muteEndpoint)
}

// SuspensionStatus handles GET /api/v1/moderation/users/{user_id}/suspension.
//
// The moderator's view of whether somebody is suspended right now.
// GET /api/v1/users/{id} carries is_active, which does read false during a
// suspension — but it reads false for a deactivated account too, and it never
// says when the suspension ends, so it cannot tell a moderator whether an early
// lift is worth offering or how much time it would actually save.
//
// Like MuteStatus and unlike the DELETE below, this does not refuse a
// self-query.
func (h *ModerationHandler) SuspensionStatus(w http.ResponseWriter, r *http.Request) {
	h.restrictionStatus(w, r, suspensionEndpoint)
}

// LiftSuspension handles DELETE /api/v1/moderation/users/{user_id}/suspension.
//
// A DELETE of the suspension for the reason LiftMute is a DELETE of the mute:
// the suspension is a value on the user (suspended_until) and this removes it.
// Modelling it as an action would imply a moderation_actions row, which the
// service deliberately does not write.
//
// It answers 204 with no body, including for a user who was not suspended.
func (h *ModerationHandler) LiftSuspension(w http.ResponseWriter, r *http.Request) {
	h.liftRestriction(w, r, suspensionEndpoint)
}

// ownPenaltyResponse is the trust cost of one action to the member it landed
// on. `decays_at` is omitted when the penalty never decays, which is readable
// as "never" rather than "unknown" only because the enclosing object is present
// at all — a member who took no penalty has no `penalty` key.
type ownPenaltyResponse struct {
	Amount   float64 `json:"amount"`
	DecaysAt string  `json:"decays_at,omitempty"`
}

// ownModerationEntryResponse is one moderation action as its subject sees it.
//
// It carries no moderator, by the same rule the mute and suspension lifts on a
// member's profile follow: which moderator handled a case appears on no
// member-facing response. In a town small enough to run this platform, naming
// the individual turns a moderation decision into a personal grievance with a
// neighbour; the decision belongs to the moderation team.
//
// service.OwnModerationEntry has already stripped the record, so this is a
// second wall rather than the only one: there is no field here for a moderator
// id, and there is none there either.
type ownModerationEntryResponse struct {
	ID        string              `json:"id"`
	Action    string              `json:"action"`
	Severity  int                 `json:"severity"`
	Reason    string              `json:"reason"`
	CreatedAt string              `json:"created_at"`
	ExpiresAt string              `json:"expires_at,omitempty"`
	Penalty   *ownPenaltyResponse `json:"penalty,omitempty"`
}

type ownModerationHistoryResponse struct {
	Actions []ownModerationEntryResponse `json:"actions"`
}

func toOwnModerationEntry(e service.OwnModerationEntry) ownModerationEntryResponse {
	resp := ownModerationEntryResponse{
		ID:        e.ID,
		Action:    string(e.Action),
		Severity:  e.Severity,
		Reason:    e.Reason,
		CreatedAt: e.CreatedAt.Format(timestampFormat),
	}
	if e.ExpiresAt != nil {
		resp.ExpiresAt = e.ExpiresAt.Format(timestampFormat)
	}
	if e.Penalty != nil {
		penalty := &ownPenaltyResponse{Amount: e.Penalty.Amount}
		if e.Penalty.DecaysAt != nil {
			penalty.DecaysAt = e.Penalty.DecaysAt.Format(timestampFormat)
		}
		resp.Penalty = penalty
	}
	return resp
}

// OwnHistory handles GET /api/v1/users/me/moderation-history.
//
// The member-facing counterpart to ListActions below, and the only moderation
// history a member has ever been able to read about themselves beyond a lifted
// mute. Before it, somebody warned for something learned that they had been
// warned only if a moderator told them out of band — the reason, the trust it
// cost and when that cost fades all sat behind the moderator-only route.
//
// It is served by the moderation handler rather than the user handler because
// the shape and the policy are moderation's, not the profile's. The handler is
// named for the subject matter; the route's guard decides the audience.
//
// The subject is the authenticated caller and is taken from the session, never
// from the URL or a query parameter. That is the whole of the authorization:
// there is no id to tamper with, so there is no way to ask this route about
// anybody else. It is also why the route needs no role floor — every member is
// entitled to their own record, and to nobody else's.
//
// It deliberately sits behind a guard that skips RequireActive; see the route
// registration in internal/server/routes.go. A suspended or banned member is
// exactly who most needs to read why.
//
// A clean record answers 200 with an empty array rather than 404: "nothing has
// happened to you" is a successful answer to the question, and the frontend has
// a reassuring empty state to render for it.
func (h *ModerationHandler) OwnHistory(w http.ResponseWriter, r *http.Request) {
	user, ok := middleware.UserFromContext(r.Context())
	if !ok {
		Error(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	entries, err := h.actions.OwnModerationHistory(
		r.Context(),
		user.ID,
		parseLimit(r.URL.Query().Get("limit")),
		parseOffset(r.URL.Query().Get("offset")),
	)
	if err != nil {
		serviceError(w, err)
		return
	}

	resp := ownModerationHistoryResponse{Actions: make([]ownModerationEntryResponse, 0, len(entries))}
	for _, e := range entries {
		resp.Actions = append(resp.Actions, toOwnModerationEntry(e))
	}

	JSON(w, http.StatusOK, resp)
}

type listActionsResponse struct {
	Actions []service.ActionHistoryEntry `json:"actions"`
}

// ListActions handles GET /api/v1/moderation/actions/{user_id}.
func (h *ModerationHandler) ListActions(w http.ResponseWriter, r *http.Request) {
	user, ok := middleware.UserFromContext(r.Context())
	if !ok {
		Error(w, http.StatusUnauthorized, "unauthorized")
		return
	}

	targetUserID := chi.URLParam(r, "user_id")
	byModerator := r.URL.Query().Get("role") == "moderator"

	if byModerator && !canQueryByModerator(user) {
		Error(w, http.StatusForbidden, "council role required")
		return
	}

	limit := parseLimit(r.URL.Query().Get("limit"))
	offset := parseOffset(r.URL.Query().Get("offset"))

	entries, err := h.actions.GetActionHistory(r.Context(), user, targetUserID, byModerator, limit, offset)
	if err != nil {
		serviceError(w, err)
		return
	}

	if entries == nil {
		entries = []service.ActionHistoryEntry{}
	}

	JSON(w, http.StatusOK, listActionsResponse{Actions: entries})
}

File: web/src/lib/moderation.ts
import type {
  ActionHistoryEntry,
  ApiError,
  MuteLift,
  MuteStatus,
  Post,
  TakeActionRequest,
  User,
} from "../api/types";
import { personName, shortId } from "./people";
import { byteLength, type ValidationResult } from "./post";

/** Mirrors the ActionType constants in internal/domain/moderation.go. */
export const ACTION_TYPES = ["warn", "mute", "suspend", "ban"] as const;

export type ActionType = (typeof ACTION_TYPES)[number];

/**
 * Mirrors allowedSeverity in internal/service/moderation_action.go. Severity is
 * bound to the action type because it is what drives the trust penalty and how
 * far through the vouch graph it propagates — a "warn" carrying ban-level
 * severity would be rejected by the server.
 */
const ALLOWED_SEVERITY: Record<ActionType, readonly number[]> = {
  warn: [1, 2],
  mute: [3],
  suspend: [4],
  ban: [5],
};

/**
 * Mirrors validateActionRequest in internal/service/moderation_action.go:
 * mutes and suspensions are temporary and must carry a duration, while bans are
 * permanent and are rejected if one is sent.
 */
const NEEDS_DURATION: readonly ActionType[] = ["mute", "suspend"];

/**
 * Everything a severity decides, mirroring the four functions over severity in
 * internal/domain/moderation.go: DirectPenalty, PenaltyDecayDays,
 * PropagationDepth and PropagationDecay.
 *
 * They are one record here rather than four lookups because the dialog needs
 * all four to describe a single action, and because a severity the server does
 * not recognize should fail once — the Go side went to some trouble to stop an
 * unknown severity being answered with a zero from each map in turn.
 */
export interface SeverityConsequence {
  /** Trust points the person acted on loses. Mirrors DirectPenalty. */
  penalty: number;
  /**
   * Days until the penalty has decayed away, or null when it never does.
   * Mirrors PenaltyDecayDays, where a recognized 0 means permanent.
   */
  decayDays: number | null;
  /** How many hops out through the vouch graph it reaches. Mirrors PropagationDepth. */
  depth: number;
  /** The fraction of the penalty surviving each hop. Mirrors PropagationDecay. */
  hopSurvival: number;
}

const SEVERITY_CONSEQUENCES: Record<number, SeverityConsequence> = {
  1: { penalty: 5, decayDays: 90, depth: 1, hopSurvival: 0.5 },
  2: { penalty: 10, decayDays: 180, depth: 1, hopSurvival: 0.7 },
  3: { penalty: 25, decayDays: 270, depth: 2, hopSurvival: 0.6 },
  4: { penalty: 40, decayDays: 365, depth: 2, hopSurvival: 0.7 },
  5: { penalty: 100, decayDays: null, depth: 3, hopSurvival: 0.75 },
};

/** The word the Go comments use for each severity, 1 through 5. */
const SEVERITY_NAMES: Record<number, string> = {
  1: "Minor",
  2: "Moderate",
  3: "Serious",
  4: "Severe",
  5: "Ban-level",
};

/**
 * What each action type does to the person it lands on, beyond the trust
 * penalty every action carries.
 *
 * Drawn from planEnforcement in internal/service/moderation_action.go and from
 * what the enforced state actually blocks: a mute is checked by
 * domain.User.CanPost alone, while a suspension is folded into is_active and so
 * meets middleware.RequireActive on every guarded route.
 */
const ACTION_EFFECTS: Record<ActionType, string> = {
  warn: "Nothing is blocked. The warning goes on their moderation record, with the reason you write below.",
  mute: "They cannot post until the mute expires. Reading the feed, reacting and vouching are unaffected.",
  suspend:
    "They cannot post, react, vouch or report until the suspension expires. Reading the feed still works.",
  ban: "Their role becomes banned and their trust score is set to zero. They cannot post or vouch again, and a ban does not expire on its own.",
};

/**
 * severityConsequence looks up what a severity costs, or null when the severity
 * is not one the server recognizes.
 *
 * Null rather than a zeroed record, for the reason the Go side returns an `ok`
 * alongside each of these: a penalty of no points, reaching nobody, decaying
 * never, is a plausible-looking answer that is wrong in every direction.
 */
export function severityConsequence(severity: number): SeverityConsequence | null {
  const c = SEVERITY_CONSEQUENCES[severity];
  // A copy, on the same principle as severitiesFor: a caller cannot edit the
  // policy by holding onto what it was told.
  return c ? { ...c } : null;
}

/**
 * Points as a moderator should read them: 28, not 28.000000000000004, which is
 * what 40 * 0.7 is in binary floating point.
 */
function formatPoints(points: number): string {
  const rounded = Math.round(points * 10) / 10;
  return Number.isInteger(rounded) ? String(rounded) : rounded.toFixed(1);
}

/**
 * How to name an action inside a sentence. A warning is named by its severity
 * because that is the one action where the moderator picks it; the other three
 * carry exactly one severity each, so naming it would only repeat the action.
 */
function actionNoun(actionType: ActionType, severity: number): string {
  switch (actionType) {
    case "warn": {
      const name = SEVERITY_NAMES[severity];
      return name ? `a ${name.toLowerCase()} warning` : "a warning";
    }
    case "mute":
      return "a mute";
    case "suspend":
      return "a suspension";
    case "ban":
      return "a ban";
  }
}

/** What taking an action will do, in the words shown to the moderator. */
export interface ActionConsequence {
  /** The action named for a sentence, e.g. "a minor warning" or "a ban". */
  noun: string;
  /** What the action does to the person it lands on. */
  effect: string;
  /** The trust penalty they take, and how long it lingers. */
  penalty: string;
  /** How far that penalty travels through the people who vouched for them. */
  propagation: string;
}

/**
 * actionConsequence describes what an action will do, so a volunteer councilor
 * is not choosing between "Level 1" and "Level 5" with nothing to tell them
 * apart.
 *
 * The propagation sentence is the reason this exists. Acting on somebody also
 * takes trust from everyone who vouched for them, out to three hops for a ban —
 * that is the whole point of a vouch graph, and it appeared nowhere in the
 * interface that applies it.
 *
 * An action type and severity the server would reject yield null rather than a
 * description, so the dialog stays silent instead of promising an outcome that
 * cannot happen. The pairing is checked against severitiesFor, which mirrors
 * allowedSeverity in internal/service/moderation_action.go.
 */
export function actionConsequence(actionType: string, severity: number): ActionConsequence | null {
  if (!severitiesFor(actionType).includes(severity)) return null;

  const c = severityConsequence(severity);
  if (!c) return null;

  const perHop = `${Math.round(c.hopSurvival * 100)}%`;
  const firstHop = formatPoints(c.penalty * c.hopSurvival);

  return {
    noun: actionNoun(actionType as ActionType, severity),
    effect: ACTION_EFFECTS[actionType as ActionType],
    penalty:
      c.decayDays === null
        ? `It costs them ${formatPoints(c.penalty)} trust points, which never decay.`
        : `It costs them ${formatPoints(c.penalty)} trust points, which decay away over ${c.decayDays} days.`,
    propagation:
      c.depth === 1
        ? `This also lowers the standing of everyone who vouched for them, one step out: each of them loses ${firstHop} points. It travels no further.`
        : `This also lowers the standing of everyone who vouched for them, up to ${c.depth} steps out through the vouch graph. Anyone who vouched for them directly loses ${firstHop} points, and each step further out carries ${perHop} of the step before it.`,
  };
}

/**
 * severityChoiceLabel names a severity in a picker, e.g. "Minor — 5 trust
 * points". "Level 1" through "Level 5" told a moderator nothing about which one
 * to choose, and the number is the input to the penalty rather than a rank
 * anybody could interpret.
 */
export function severityChoiceLabel(severity: number): string {
  const c = severityConsequence(severity);
  const name = SEVERITY_NAMES[severity] ?? `Level ${severity}`;
  return c ? `${name} — ${formatPoints(c.penalty)} trust points` : name;
}

/** Mirrors maxActionReasonLen in internal/service/moderation_action.go. */
export const MAX_ACTION_REASON_LENGTH = 1000;

/** Mirrors maxRemovalReasonLen in internal/service/post.go. */
export const MAX_REMOVAL_REASON_LENGTH = 1000;

/**
 * validateRemovalReason applies the server's rules for a moderator's post
 * removal note so the dialog can disable its submit button instead of
 * round-tripping to a guaranteed 400.
 *
 * A reason is mandatory here, unlike an author deleting their own post: it is
 * the only record of why a moderator overrode a member's speech.
 *
 * The length is measured in UTF-8 bytes because the server's check is
 * `len(reason)` on a Go string. Measuring JS string length would wave through a
 * reason of multi-byte characters that the server then rejects.
 */
export function validateRemovalReason(reason: string): ValidationResult {
  const trimmed = (reason ?? "").trim();
  if (trimmed.length === 0) {
    return { valid: false, error: "A reason is required." };
  }
  const bytes = byteLength(trimmed);
  if (bytes > MAX_REMOVAL_REASON_LENGTH) {
    return {
      valid: false,
      error: `Reason is ${bytes} bytes; the maximum is ${MAX_REMOVAL_REASON_LENGTH}.`,
    };
  }
  return { valid: true };
}

/**
 * canRemovePost reports whether the queue should offer to take a post down.
 *
 * Mirrors canRemoveByModerator in internal/service/post.go: only a visible post
 * may be removed, so offering the control on one that is already gone would
 * guarantee a 400. A null post is one the card has not loaded yet, or failed
 * to — either way there is nothing to act on.
 */
export function canRemovePost(post: Post | null | undefined): boolean {
  return post?.status === "visible";
}

/**
 * describeHistorySubject names the member whose moderation history is on
 * screen, for the heading of that page.
 *
 * There is no read that hands the page a member's profile — it loads a list of
 * actions and nothing else — so the name comes from the actions themselves,
 * every one of which is addressed to this member and carries
 * `target_display_name`. Any entry will do; the first one with a name wins, so
 * a member named on some rows and not others is still named.
 *
 * This is only sound for the default listing, which is actions taken *against*
 * the user. The `role=moderator` variant lists actions they took, where the
 * subject is the moderator rather than the target — this page does not ask for
 * that variant, and would need the other field if it did.
 *
 * The short id stays alongside the name rather than being replaced by it. Two
 * members may share a display name, and a moderator about to act on somebody
 * needs to know which of them they are reading. When there is no name it is all
 * there is to show, and it is not repeated.
 */
export function describeHistorySubject(
  entries: readonly ActionHistoryEntry[] | null | undefined,
  userId: string,
): string {
  const named = entries?.find((entry) => entry?.action?.target_display_name?.trim());
  const name = personName(named?.action?.target_display_name, userId);
  const short = shortId(userId);

  return name === short ? short : `${name} · ${short}`;
}

/** What the queue should do with a report after acting on it. */
export interface ResolutionResult {
  /** Whether the report may be dropped from the queue. */
  resolved: boolean;
  /** Why it could not be, present only when `resolved` is false. */
  error?: string;
}

/**
 * reportResolutionOutcome decides what the queue does after it has tried to
 * mark a report reviewed.
 *
 * Taking action used to drop the report from local state and never tell the
 * server, so it came back `pending` on the next load and moderators re-reviewed
 * work they had already done. Resolving it is a second request, and this decides
 * what happens when that request is the thing that fails.
 *
 * A 404 still resolves: another moderator got there first, so the report is
 * gone either way and holding it in this moderator's queue helps nobody. This
 * matches how ReportCard already treats a 404 when dismissing.
 *
 * Any other failure does NOT resolve. The action or removal already succeeded,
 * so the report is genuinely still pending on the server — dropping it from the
 * queue anyway is exactly the divergence this function exists to stop, and the
 * message says the action stuck so the moderator does not redo it.
 */
export function reportResolutionOutcome(err: ApiError | null): ResolutionResult {
  if (err === null) {
    return { resolved: true };
  }
  if (err.status === 404) {
    return { resolved: true };
  }
  const detail = err.error ? `: ${err.error}` : ".";
  return {
    resolved: false,
    error: `The action went through, but the report could not be marked reviewed${detail} It is still in the queue.`,
  };
}

/** The largest duration the dialog offers, one year in hours. */
export const MAX_DURATION_HOURS = 8760;

const SECONDS_PER_HOUR = 3600;

function isActionType(value: string): value is ActionType {
  return (ACTION_TYPES as readonly string[]).includes(value);
}

/**
 * severitiesFor lists the severities valid for an action type, in the order the
 * picker should offer them — the first is the one to preselect.
 *
 * An unknown action type yields an empty list rather than throwing, so a server
 * that starts sending a new action type cannot break the dialog.
 */
export function severitiesFor(actionType: string): number[] {
  return isActionType(actionType) ? [...ALLOWED_SEVERITY[actionType]] : [];
}

/** needsDuration reports whether an action type is a temporary one. */
export function needsDuration(actionType: string): boolean {
  return isActionType(actionType) && NEEDS_DURATION.includes(actionType);
}

export interface ActionInput {
  actionType: string;
  severity: number;
  reason: string;
  /** The signed-in moderator taking the action. */
  moderatorId: string;
  /** The member the action lands on. */
  targetUserId: string;
}

/**
 * validateAction applies the server's rules so the dialog can disable its
 * submit button instead of round-tripping to a guaranteed 400.
 *
 * durationHours is only consulted for the action types that need one; passing
 * it is optional so callers that have not collected a duration yet can still
 * check the rest of the form.
 *
 * The two ids are part of the input rather than extra parameters because
 * buildActionRequest needs the target too, and one struct means the request
 * cannot be built for a different person than the one just validated.
 */
export function validateAction(input: ActionInput, durationHours?: number): ValidationResult {
  // Checked before every field, and deliberately earlier than the server checks
  // it — validateActionRequest in internal/service/moderation_action.go
  // validates type, severity and reason first. Those are all things the
  // moderator can fix in this form; acting on themselves is not, so reporting a
  // severity or reason problem first would send them to correct fields that
  // were never the reason the action could not go through. It is also the more
  // specific answer, the same reasoning that puts the self-check ahead of the
  // trust gate in vouchBlockReason.
  //
  // Both ids must be present to count as a match: a viewer whose identity has
  // not loaded yet has two empty strings, and that is unknown, not self.
  if (input?.moderatorId && input.moderatorId === input?.targetUserId) {
    return { valid: false, error: "You cannot moderate yourself." };
  }

  if (!input || !isActionType(input.actionType)) {
    return { valid: false, error: "Select an action type." };
  }

  if (!ALLOWED_SEVERITY[input.actionType].includes(input.severity)) {
    return {
      valid: false,
      error: `Severity ${input.severity} is not valid for a ${input.actionType}.`,
    };
  }

  const reason = (input.reason ?? "").trim();
  if (reason.length === 0) {
    return { valid: false, error: "A reason is required." };
  }
  if (reason.length > MAX_ACTION_REASON_LENGTH) {
    return {
      valid: false,
      error: `Reason is ${reason.length} characters; the maximum is ${MAX_ACTION_REASON_LENGTH}.`,
    };
  }

  if (needsDuration(input.actionType)) {
    if (!Number.isFinite(durationHours) || (durationHours as number) <= 0) {
      return { valid: false, error: `A ${input.actionType} needs a duration in hours.` };
    }
    if ((durationHours as number) > MAX_DURATION_HOURS) {
      return {
        valid: false,
        error: `Duration must be at most ${MAX_DURATION_HOURS} hours.`,
      };
    }
  }

  return { valid: true };
}

/**
 * buildActionRequest assembles the request body.
 *
 * The target is read off the same input validateAction was given rather than
 * passed separately, so the request cannot be addressed to someone other than
 * the person the self-moderation check cleared.
 *
 * The duration is omitted entirely for action types that do not take one: the
 * server rejects a ban that carries a duration, so sending a stale value left
 * in the form field would fail the whole action.
 */
export function buildActionRequest(
  input: ActionInput,
  durationHours?: number,
): TakeActionRequest {
  return {
    target_user_id: input?.targetUserId ?? "",
    action_type: input.actionType,
    severity: input.severity,
    reason: (input.reason ?? "").trim(),
    duration_seconds: needsDuration(input.actionType)
      ? Math.round((durationHours as number) * SECONDS_PER_HOUR)
      : undefined,
  };
}

/**
 * activeMuteExpiry reads GET /api/v1/moderation/users/{id}/mute and reports
 * when the mute ends, or null when there is no mute in force.
 *
 * The server already omits muted_until for a user who is not muted and for one
 * whose mute has expired, so an absent field is the answer to "is this person
 * muted?" — the same rule the caller's own profile uses. The expiry is
 * re-checked against `now` anyway, because a page left open outlives the
 * response that loaded it and would otherwise keep offering to lift a mute that
 * has since run out on its own.
 *
 * A timestamp that cannot be parsed reads as no mute rather than an Invalid
 * Date, matching formatDateTime's contract: the moderator sees the plain view
 * instead of a control claiming to lift something unnamed.
 */
export function activeMuteExpiry(
  status: MuteStatus | null | undefined,
  now: Date,
): Date | null {
  return parseActiveExpiry(status?.muted_until, now);
}

/**
 * liftMuteBlockReason mirrors canLiftMute in
 * internal/service/moderation_action.go so the page can explain itself instead
 * of round-tripping to a guaranteed 400.
 *
 * A moderator may not lift a mute placed on themselves: that is the one case no
 * route guard can catch, since a muted moderator satisfies every middleware in
 * the chain. Both ids must be present to count as a match — a viewer whose
 * identity has not loaded yet has two empty strings, and that is unknown, not
 * self, exactly as validateAction treats it.
 */
export function liftMuteBlockReason(viewerId: string, targetId: string): string | null {
  if (viewerId && viewerId === targetId) {
    return "You cannot moderate yourself.";
  }
  return null;
}

/**
 * What a member's own profile has to say about their moderation state.
 *
 * A mute and a lift are separate facts and both can be true at once: being
 * muted again does not undo having been released from an earlier mute, so
 * collapsing them into one status would hide half the record.
 */
export interface OwnMuteNotice {
  /** When an active mute ends, or null when none is in force. */
  mutedUntil: Date | null;
  /** The most recent early release, or null when there has never been one. */
  latestLift: MuteLift | null;
  /** Whether there is anything at all to show, so the caller can skip the UI. */
  hasAnything: boolean;
}

/**
 * ownMuteNotice reads the moderation half of the caller's own profile.
 *
 * It exists because muted_until disappears the moment a mute is lifted. Without
 * the lift record a member released early would see their profile simply stop
 * saying they were muted, with nothing anywhere to say it happened — the whole
 * moderation audit trail sits behind the moderator-only /v1/moderation routes.
 *
 * The expiry is re-checked against `now` even though the server already omits
 * an expired muted_until: a page left open outlives the response that loaded
 * it, and would otherwise go on telling a member they cannot post. An
 * unparseable timestamp reads as no mute rather than an Invalid Date, matching
 * activeMuteExpiry.
 *
 * Another member's profile carries neither field, so it reports nothing — this
 * is never a way to learn about somebody else's moderation.
 */
export function ownMuteNotice(user: User | null | undefined, now: Date): OwnMuteNotice {
  const mutedUntil = parseActiveExpiry(user?.muted_until, now);
  const latestLift = user?.mute_lifts?.[0] ?? null;

  return {
    mutedUntil,
    latestLift,
    hasAnything: mutedUntil !== null || latestLift !== null,
  };
}

/** Shared by ownMuteNotice and activeMuteExpiry: a timestamp still in the future, or null. */
function parseActiveExpiry(raw: string | undefined, now: Date): Date | null {
  if (!raw) return null;

  const expiry = new Date(raw);
  if (Number.isNaN(expiry.getTime())) return null;

  return expiry.getTime() > now.getTime() ? expiry : null;
}
